src/realm/object-store/sync/sync_manager.cpp

```diff
@@ -446,6 +446,57 @@ void SyncManager::remove_user(const std::string& user_id)
     }
 }
 
+void SyncManager::close()
+{
+    std::lock_guard<std::mutex> lock(m_file_system_mutex);
+    m_metadata_manager = nullptr;
+    m_client_uuid = util::none;
+    m_file_manager = nullptr;
+
+    {
+        // Destroy all the users.
+        std::lock_guard<std::mutex> lock(m_user_mutex);
+        m_users.clear();
+        m_current_user = nullptr;
+    }
+
+    std::vector<std::shared_ptr<SyncSession>> current_sessions;
+
+    {
+        std::lock_guard<std::mutex> lock(m_session_mutex);
+
+        for (auto& session : m_sessions) {
+            current_sessions.push_back(session.second);
+        }
+    }
+
+    for (auto& session : current_sessions) {
+        session->shutdown_and_wait();
+    }
+
+    {
+        std::lock_guard<std::mutex> lock(m_session_mutex);
+        m_sessions.clear();
+        current_sessions.clear();
+    }
+
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+
+        // Stop the client. This will abort any uploads that inactive sessions are waiting for.
+        if (m_sync_client)
+            m_sync_client->stop();
+
+        // Destroy the client now that we have no remaining sessions.
+        m_sync_client = nullptr;
+
+        // Reset even more state.
+        m_config = {};
+
+        m_sync_route = "";
+    }
+}
```

Comment:
I think this entire method can be simplified to just stopping the sync client and forcibly shutting down the sessions.
```cpp
if (m_sync_client)
     m_sync_client->stop();
for (auto& session : current_sessions) {
     session->shutdown_and_wait();
}
```
All of the in memory nullification you're doing should naturally be happening. I see that you left a comment about why we need to use this method and reset all that state, but– we should probably just fix the reference loop between users and sync managers. There is a (semi)clear hierarchy here– an App has a SyncManager and Users. If there is no App, there should not be a sync manager nor users. The User should be able to have only a weak reference to the SyncManager.

Reply:
The memory nullification is done because we're not in a destructor, so won't happen automatically. If we resolved the reference cycle and moved it to a destructor, then we can remove most of what's there.

The issue with keeping a weak pointer to the sync manager on the user manifests itself precisely during session termination. When the session moves to inactive state, it will attempt to look up the sync manager from the user and unregister itself. If the sync manager is being destroyed, the user won't be able to lock the weak pointer and thus the session won't be able to call unregister.

One approach would be not to proactively unregister inactive sessions from the manager and instead prune them whenever sessions is accessed, but that was a larger code change than I was willing to make, considering all of that code is virtually a black box to me 😅